### PR TITLE
KEYCLOAK-19138 nginx x509 client certificate provider trusting nginx …

### PR DESCRIPTION
**Pull request by yelhouti**
…validation

### CODE REVIEW

**Comment by pedroigor** [comment]
@yelhouti Thanks and sorry for the delay. 

I understand the reasons behind the changes as per https://issues.redhat.com/browse/KEYCLOAK-19138. By relying on the validation at the NGINX side we might not need to resolve all the certificate chain.

I would rather add a new provider instead of changing the existing one. Are you still willing to contribute to the change?

**Comment by yelhouti** [comment]
@pedroigor no worries, of course I am willing to contribute if you can just tell me the name you want for the new provider :)

**Comment by pedroigor** [comment]
@yelhouti Actually, we don't need a new provider factory implementation.

What do you think about:

* Support a `trust-client-verification` configuration option (naming is hard, anything else if you have a better one)
* Create a `NginxProxyTrustedClientCertificateLookup`based on the implementation you did here
* When creating the provider from the factory's `create` method, return an instance of the above class if the new option is set.

**Comment by yelhouti** [comment]
@pedroigor I couldn't test everything, the new console fails when trying to create a X509/Validate Username Form

File: services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java [hunk 1]
```diff
@@ -55,6 +57,8 @@ public void init(Config.Scope config) {
             } else {
                 logger.tracev("{0} was not configured", HTTP_HEADER_CERT_CHAIN_PREFIX);
             }
+            trustProxyVerification = config.getBoolean(TRUST_PROXY_VERIFICATION, false);
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java:60]: Does it make more sense to have this option only available to nginx factory?

**Review by pedroigor — COMMENTED** [review]
@yelhouti Thanks a lot. Only a single comment.

File: services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java (hunk 1 above)
Comment (yelhouti) [reply services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java:60]: I guess it is true by default in the others, but some one might choose to have a trusted and non trusted apache and use the same variable.

Comment (yelhouti) [reply services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java:60]: If we want to do that, I think I would need to override the init, to call super.init then store the variable. I can do that if you want.

Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java:60]: Yeah, I see your point. But for now, I would keep it only available to nginx and if needed, refactor it later.

Sorry for asking this, you have been waiting for a long time already :) But just this one and we are good to go.

Comment (yelhouti) [reply services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java:60]: no worries, on it

Comment (yelhouti) [reply services/src/main/java/org/keycloak/services/x509/AbstractClientCertificateFromHttpHeadersLookupFactory.java]: @pedroigor here you go

**Review by pedroigor — DISMISSED** [review]

File: services/src/main/java/org/keycloak/services/x509/NginxProxySslClientCertificateLookupFactory.java [hunk 2]
```diff
@@ -13,12 +15,30 @@
...
+        super.init(config);
+        trustProxyVerification = config.getBoolean(TRUST_PROXY_VERIFICATION, false);
+        logger.tracev("{0}: ''{1}''", TRUST_PROXY_VERIFICATION, trustProxyVerification);
+    }
+
     @Override
     public X509ClientCertificateLookup create(KeycloakSession session) {
-        return new NginxProxySslClientCertificateLookup(sslClientCertHttpHeader,
-                sslChainHttpHeaderPrefix, certificateChainLength, session);
+        if (trustProxyVerification) {
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/x509/NginxProxySslClientCertificateLookupFactory.java:35]: Humm ... Looks like the logic is inverted? We should return NginxProxyTrustedClientCertificateLookup if the options is set to `true`, right?

Comment (yelhouti) [reply services/src/main/java/org/keycloak/services/x509/NginxProxySslClientCertificateLookupFactory.java:35]: @pedroigor you saw nothing :p, no it's not :p

**Review by pedroigor — APPROVED** [review]

**Comment by yelhouti** [comment]
@pedroigor, could you merge it please :)

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@yelhouti Thanks.

**Comment by yelhouti** [comment]
@pedroigor thank you, any idea on when the next release will be please ?

**Comment by pedroigor** [comment]
Targeted to 20.

**Pull request merged**